Source/WebCore/rendering/svg/SVGContainerLayout.cpp

```diff
@@ -75,19 +76,20 @@ void SVGContainerLayout::layoutChildren(bool containerNeedsLayout)
             if (child.isAnonymous()) {
                 ASSERT(is<RenderSVGViewportContainer>(child));
                 needsLayout = true;
-            } else if (auto* element = dynamicDowncast<SVGElement>(*child.node())) {
+            } else if (auto* element = dynamicDowncast<SVGElement>(*child.node()); element && element->hasRelativeLengths()) {
                 // When containerNeedsLayout is false and the layout size changed, we have to check whether this child uses relative lengths
-                if (element->hasRelativeLengths()) {
-                    // When the layout size changed and when using relative values tell the RenderSVGShape to update its shape object
-                    if (CheckedPtr shape = dynamicDowncast<RenderSVGShape>(child))
-                        shape->setNeedsShapeUpdate();
-                    else if (CheckedPtr svgText = dynamicDowncast<RenderSVGText>(child)) {
-                        svgText->setNeedsTextMetricsUpdate();
-                        svgText->setNeedsPositioningValuesUpdate();
-                    }
 
+                // When the layout size changed and when using relative values tell the RenderSVGShape to update its shape object
+                if (CheckedPtr shape = dynamicDowncast<RenderSVGShape>(child)) {
+                    shape->setNeedsShapeUpdate();
                     needsLayout = true;
-                }
+                } else if (CheckedPtr svgText = dynamicDowncast<RenderSVGText>(child)) {
+                    svgText->setNeedsTextMetricsUpdate();
+                    svgText->setNeedsPositioningValuesUpdate();
+                    needsLayout = true;
+                } else if (CheckedPtr resource = dynamicDowncast<RenderSVGResourceGradient>(child))
+                    resource->invalidateGradient();
```

Comment:
This change was necessary to fix `LayoutTests/svg/custom/relative-sized-content-with-resources.xhtml` - if a gradient uses relative lengths for e.g. cx/cy/fx/etc. (for radial gradients) it needs to recompute the internal `Gradient` and repaint all clients. In LBSE we can easily solve that during `SVGContainerLayout::layoutChildren` - without resorting to artificial relayout, as the legacy engine did.

Reply:
It seem the layout and repaint phases get intermingled this way, but I guess LayoutRepainter is a bit similar and maybe there are other cases like this....